lib/srv/desktop/rdp/rdpclient/src/rdpdr/filesystem.rs

```diff
@@ -768,100 +753,116 @@ impl std::fmt::Display for FilesystemBackendError {
+
+    fn new(
+        handler: impl FnOnce(&mut FilesystemBackend, Self::ResponseType) -> PduResult<()>
+            + Send
+            + 'static,
+    ) -> Self;
+
+    fn call(self, this: &mut FilesystemBackend, res: Self::ResponseType) -> PduResult<()>;
+}
 
+/// See the example in [`ResponseHandler`].
+macro_rules! response_handler {
+    ($name:ident, $response_type:ty) => {
         struct $name {
-            handler: $func_name,
+            handler:
+                Box<dyn FnOnce(&mut FilesystemBackend, $response_type) -> PduResult<()> + Send>,
         }
 
-        impl $name {
-            fn new(handler: $func_name) -> Self {
-                Self { handler }
+        impl std::fmt::Debug for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                write!(f, stringify!($name))
             }
+        }
 
-            fn call(
-                self,
-                this: &mut FilesystemBackend,
-                res: $response_type,
-            ) -> PduResult<Option<RdpdrPdu>> {
-                (self.handler)(this, res)
+        impl ResponseHandler for $name {
+            type ResponseType = $response_type;
+
+            fn new(
+                handler: impl FnOnce(&mut FilesystemBackend, Self::ResponseType) -> PduResult<()>
+                    + Send
+                    + 'static,
+            ) -> Self {
+                Self {
+                    handler: Box::new(handler),
+                }
             }
-        }
 
-        impl std::fmt::Debug for $name {
-            fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-                write!(f, concat!("<", stringify!($name), ">"))
+            fn call(self, this: &mut FilesystemBackend, res: Self::ResponseType) -> PduResult<()> {
+                (self.handler)(this, res)
             }
         }
     };
 }
 
 response_handler!(
     SharedDirectoryInfoResponseHandler,
-    InfoResponseHandlerFunction,
     tdp::SharedDirectoryInfoResponse
 );
-
 response_handler!(
     SharedDirectoryCreateResponseHandler,
-    CreateResponseHandlerFunction,
     tdp::SharedDirectoryCreateResponse
 );
-
 response_handler!(
     SharedDirectoryDeleteResponseHandler,
-    DeleteResponseHandlerFunction,
     tdp::SharedDirectoryDeleteResponse
 );
+
+type CompletionId = u32;
+
+/// A generic cache for storing [`ResponseHandler`]s indexed by [`CompletionId`].
+///
+/// See the example in [`ResponseHandler`].
+#[derive(Debug)]
+struct ResponseCache<T: ResponseHandler> {
+    cache: HashMap<CompletionId, T>,
+}
+
+impl<T: ResponseHandler> ResponseCache<T> {
+    fn new() -> Self {
+        Self {
+            cache: HashMap::new(),
+        }
+    }
+
+    fn insert(&mut self, completion_id: CompletionId, handler: T) {
+        self.cache.insert(completion_id, handler);
+    }
+
+    fn remove(&mut self, completion_id: &CompletionId) -> Option<T> {
+        self.cache.remove(completion_id)
+    }
+}
```

Comment:
Moving https://github.com/gravitational/teleport/pull/34011#discussion_r1376829352 to here.

Yes that's a better idea, see: c9913e7b1bde102a6649c98d818636df8a6db093